src/htm/regions/TMRegion.cpp

```diff
@@ -239,24 +241,24 @@ void TMRegion::compute() {
     NTA_DEBUG << "bottomUpOut " << *out << std::endl;
   }
   out = getOutput("activeCells");
-  if (out && (out->hasOutgoingLinks() || LogItem::isDebug())) {
+  if (out && out->hasOutgoingLinks() ) {
     tm_->getActiveCells(out->getData().getSDR());
     NTA_DEBUG << "active " << *out << std::endl;
   }
   out = getOutput("predictedActiveCells");
-  if (out && (out->hasOutgoingLinks() || LogItem::isDebug())) {
+  if (out && out->hasOutgoingLinks() ) {
```

Comment:
@dkeeney this is something I'd like to consult with you. Actually the problem I'm hitting here seems to be a bug in TMRegion, 
a) I'm removing the "is Debug" part of the check, as it was supposed to be a helper for debugging, but breaks the logic (tests expecting such behavior are wrong) 
b) there's a problem with requiring the `hasOutgoingLinks()`, as: 

This is the test failing: 
```
net = engine.Network()
#net.setLogLevel(htm.bindings.engine_internal.LogLevel.Verbose)     # Verbose shows data inputs and outputs while executing.

encoder = net.addRegion("encoder", "ScalarSensor", "{n: 6, w: 2}");
sp = net.addRegion("sp", "SPRegion", "{columnCount: 200}");
tm = net.addRegion("tm", "TMRegion", "");
net.link("encoder", "sp");
net.link("sp", "tm");
net.initialize();

encoder.setParameterReal64("sensedValue", 0.8);  #Note: default range setting is -1.0 to +1.0
net.run(1)

sp_input = sp.getInputArray("bottomUpIn")
sdr = sp_input.getSDR()
self.assertTrue(np.array_equal(sdr.sparse, EXPECTED_RESULT1))

sp_output = sp.getOutputArray("bottomUpOut")
sdr = sp_output.getSDR()
self.assertTrue(np.array_equal(sdr.sparse, EXPECTED_RESULT2))

tm_output = tm.getOutputArray("predictedActiveCells")
sdr = tm_output.getSDR()
self.assertTrue(np.array_equal(sdr.sparse, EXPECTED_RESULT3))
```

The network is `Encoder->SP->TM`, 
I was surprised only the TM check is failing, but it is because the TM is a "leaf" (it's last part of the Network), and we have the condition `if (out && out->hasOutgoingLinks() ) `. 
now, we want a similar check (to "has outputs") so that we only compute for connected regions. Using `getInput("name").hasIncomingLinks()` would work for TM in this example, but would fail for the root (sensor/encoder). 

So do we want a combination and determine if XXX (what is the "predictedActiveCells" called, it's not a region per se...?) is connected by 
```
out = getOutput("predictedActiveCells");
if ((out && out->hasOutgoingLinks()) || (getInput("predictedActiveCells") &&  getInput("predictedActiveCells")->getIncomingLinks() ) ) { //means  either XX,or YY exists and is connected in a chain XX->TM->YY
tm_->activateDendrites();
tm_->getWinnerCells(out->getData().getSDR());
NTA_DEBUG << "winners " << *out << std::endl;
}
```

Reply:
I am having a problem understanding your post on my cell phone.
I will have to check it out later when I return.

Reply:
> ` if (out && (out->hasOutgoingLinks() || LogItem::isDebug())) {`

This logic is correct.  Although we need someplace else to put isDebug() since you removed LogItem.   This is part of the trace facility.  When the debug mode is activated by an application by calling setLogLevel(LogLevel_Verbose), the trace should show each region that is executed and the outputs.

For the trace facility to work, we need at least the following:
- LogtLevel must be global.
- We need a function like setLogLevel( ) for applications to call.  (they should not have access to the global LogLevel variable).  
- We need a function like isDebug( ) to allow conditional processing related to the trace.
- We need a function (or Macro) like NTA_DEBUG( ) to indicate what is to be displayed in the trace.

This trace facility is extremely useful when debugging a problem with region execution and or linking.   In particular, which data is being passed at which times and the order that the regions are executed.   This is another case where we need more documentation and explanation of the 'features' that are available.

Oh, and the 'Verbose' facility that I used with some of the Unit Tests does not use the trace facility.

TMRegion has several  'optional' outputs.  For the 'optional' outputs, if there is no outgoing connection, then the output is not generated.  So, to get the trace in Verbose mode we should generate the output data even though it is not being sent to an output.   

However, recently we exposed the ability to access the output buffers directly with region.getOutput(name).getData().  This causes a problem because if there is no outgoing link the optional outputs cannot be accessed in this way.  Perhaps to be on the safe side we should always produce all outputs even if they are not used.  In that case we can remove isDebug( ) and remove the entire 'if' statement at the top but the cost is lower performance when the outputs are not needed.

Reply:
> So do we want a combination and determine if XXX (what is the "predictedActiveCells" called, it's not a region per se...?) is connected by ....

I have no idea what you are trying to say here.   Could you re-phrase it?

Reply:
>>   if (out && (out->hasOutgoingLinks() || LogItem::isDebug())) {

> This logic is correct. Although we need someplace else to put isDebug() since you removed LogItem.

this is what I wanted to disprove. Debug is useful but we should not change regions' output based on it (as the output is what tests check)

> For the trace facility to work, we need at least the following:

we have these, as: 

>    LogtLevel must be global.

`NTA_LOG_LEVEL`

>    We need a function like setLogLevel( ) for applications to call. (they should not have access to the global LogLevel variable).

`Network.setLogLevel()` (but they do have access to the global variable) 

>    We need a function like isDebug( ) to allow conditional processing related to the trace.

`NTA_LOG_LEVEL == htm::LogLevel::LogLevel_Verbose`

>    We need a function (or Macro) like NTA_DEBUG( ) to indicate what is to be displayed in the trace.

can we use NTA_DEBUG? 

> TMRegion has several 'optional' outputs. For the 'optional' outputs, if there is no outgoing connection, then the output is not generated. So, to get the trace in Verbose mode we should generate the output data even though it is not being sent to an output. That is what the above 'if' statement does.

going back to what I was trying to prove wrong. 

> For the 'optional' outputs, if there is no outgoing connection, then the output is not generated

and 

> However, recently we exposed the ability to access the output buffers directly with region.getOutput(name).getData(). This causes a problem because if there is no outgoing link the optional outputs cannot be accessed in this way. Perhaps to be on the safe side we should always produce all outputs even if they are not used. 

I'm not sure what the "optional" output is in this context, but this is the bug. In the test mentioned, the Network looks like `encoder -> SP -> TM` with TM's `getOutput("predictedActiveCells")` apparently being optional. But users (and our tests) can expect the output and query for it, and it's empty. 

So we should either: 
- always produce all outputs, as you suggest.
- or document this and require user to mark the (optional) output as required. How would one do that? (eg on network_test.py, see https://github.com/htm-community/htm.core/pull/742#discussion_r341826453 )
- what I wanted to do in 2149e83 is extending the limitation from "has outgoing" to "has in or out links". If you look at it as a graph problem, a network with (some) disconnected regions should not compute them (`E   A->B->C` : do not compute E). Now we compute all of `B` because it has out link to C. But we do not compute optional outputs of `C` (as it has no outgoing links)

Reply:
> Or we make a big step and deprecate the macros and use the Log class

Oh, I don't propose deprecating the macros, just change the macros to use the Log class.

If you declare the class as a static variable in the Log.hpp file it will create a separate instance of it in each .cpp in which Log.hpp is included.  But that is ok since we only need state from the class variable which will be global in scope.  An alternative is to go head and use the thread global space to store the LogLevel state (prefered).  Either way we would not need a Log.cpp.  Everything could be hidden by Macros so that they can be changed without changing the API.